Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm

```diff
@@ -0,0 +1,569 @@
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/action/getBadgeText
+
+    std::optional<WebExtensionWindowIdentifier> windowIdentifier;
+    std::optional<WebExtensionTabIdentifier> tabIdentifier;
+    if (!parseActionDetails(details, windowIdentifier, tabIdentifier, outExceptionString))
+        return;
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ActionGetBadgeText(windowIdentifier, tabIdentifier), [protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<String> badgeText, std::optional<String> error) {
+        if (error) {
+            callback->reportError(error.value());
+            return;
+        }
+
+        ASSERT(badgeText);
+
+        callback->call((NSString *)badgeText.value());
+    }, extensionContext().identifier().toUInt64());
+}
+
+void WebExtensionAPIAction::setBadgeText(NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/action/setBadgeText
+
+    static NSArray<NSString *> *requiredKeys = @[ textKey ];
+
+    static NSDictionary<NSString *, id> *types = @{
+        textKey: [NSOrderedSet orderedSetWithObjects:NSString.class, NSNull.class, nil],
+    };
+
+    if (!validateDictionary(details, @"details", requiredKeys, types, outExceptionString))
+        return;
+
+    std::optional<WebExtensionWindowIdentifier> windowIdentifier;
+    std::optional<WebExtensionTabIdentifier> tabIdentifier;
+    if (!parseActionDetails(details, windowIdentifier, tabIdentifier, outExceptionString))
+        return;
+
+    String text = nullString();
+    if (NSString *string = objectForKey<NSString>(details, textKey, false))
+        text = string;
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ActionSetBadgeText(windowIdentifier, tabIdentifier, text), [protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<String> error) {
+        if (error) {
+            callback->reportError(error.value());
+            return;
+        }
+
+        callback->call();
+    }, extensionContext().identifier().toUInt64());
+}
+
+void WebExtensionAPIAction::getBadgeBackgroundColor(NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/action/getBadgeBackgroundColor
+
+    std::optional<WebExtensionWindowIdentifier> windowIdentifier;
+    std::optional<WebExtensionTabIdentifier> tabIdentifier;
+    if (!parseActionDetails(details, windowIdentifier, tabIdentifier, outExceptionString))
+        return;
+
+    // FIXME: <rdar://problem/57666368> Implement getting/setting the extension toolbar item's badge background color.
+
+    callback->call(@[ @255, @0, @0, @255 ]);
```

Comment:
Oh I see - you implemented the badge background color APIs as stubs.

Reply:
Yeah, this matches Safari's implementations.